fpmsyncd/routesync.cpp

```diff
@@ -756,88 +790,143 @@ void RouteSync::onRouteMsg(int nlmsg_type, struct nl_object *obj, char *vrf)
             return;
     }
 
-    struct nl_list_head *nhs = rtnl_route_get_nexthops(route_obj);
-    if (!nhs)
-    {
-        SWSS_LOG_INFO("Nexthop list is empty for %s", destipprefix);
-        return;
-    }
-
-    /* Get nexthop lists */
+    vector<FieldValueTuple> fvVector;
     string gw_list;
     string intf_list;
     string mpls_list;
-    getNextHopList(route_obj, gw_list, mpls_list, intf_list);
-    string weights = getNextHopWt(route_obj);
 
-    vector<string> alsv = tokenize(intf_list, NHG_DELIMITER);
-    for (auto alias : alsv)
+#ifdef HAVE_NEXTHOP_GROUP
+    string nhg_id_key;
+    uint32_t nhg_id = rtnl_route_get_nh_id(route_obj);
+    if(nhg_id)
     {
-        /*
-         * An FRR behavior change from 7.2 to 7.5 makes FRR update default route to eth0 in interface
-         * up/down events. Skipping routes to eth0 or docker0 to avoid such behavior
-         */
-        if (alias == "eth0" || alias == "docker0")
+        const auto itg = m_nh_groups.find(nhg_id);
+        if(itg == m_nh_groups.end())
+        {
+            SWSS_LOG_DEBUG("NextHop group id %d not found. Dropping the route %s", nhg_id, destipprefix);
+            return;
+        }
+        NextHopGroup& nhg = itg->second;
+        if(nhg.group.size() == 0)
+        {
+            //Using route-table only for single next-hop
+            string nexthops, ifnames, weights;
+
+            getNextHopGroupFields(nhg, nexthops, ifnames, weights, rtnl_route_get_family(route_obj));
+            FieldValueTuple gw("nexthop", nexthops.c_str());
+            FieldValueTuple intf("ifname", ifnames.c_str());
+            fvVector.push_back(gw);
+            fvVector.push_back(intf);
+            SWSS_LOG_DEBUG("NextHop group id %d is a single nexthop address. Filling the route table %s with nexthop and ifname", nhg_id, destipprefix);
+        }
+        else
+        {
+            nhg_id_key = getNextHopGroupKeyAsString(nhg_id);
+            FieldValueTuple nhg("nexthop_group", nhg_id_key.c_str());
+            fvVector.push_back(nhg);
+            updateNextHopGroup(nhg_id);
+        }
+
+        auto proto_num = rtnl_route_get_protocol(route_obj);
+        auto proto_str = getProtocolString(proto_num);
+        FieldValueTuple proto("protocol", proto_str);
+        fvVector.push_back(proto);
+
+    }
+    else
+#endif
+    {
+        struct nl_list_head *nhs = rtnl_route_get_nexthops(route_obj);
+        if (!nhs)
+        {
+            SWSS_LOG_INFO("Nexthop list is empty for %s", destipprefix);
+            return;
+        }
+
+        /* Get nexthop lists */
+
+        getNextHopList(route_obj, gw_list, mpls_list, intf_list);
+        string weights = getNextHopWt(route_obj);
+
+        vector<string> alsv = tokenize(intf_list, NHG_DELIMITER);
+        for (auto alias : alsv)
```

Comment:
What is the purpose of this loop? Is it to print the skipped routes ? Because the only logic there is when the list is of size 1.

Reply:
Thanks, I moved if(alsv.size() == 1)) outside the loop.